### ISSUE TITLE
Exotic fetching type with associations and database functions

### ISSUE THREAD

**Issue by ManheimGreen** [groue/GRDB.swift#1013]
Hello. GRDB is a work of genius. (As everyone knows.)

Unfortunately, I’m not a genius with database in general, and I’m having trouble with a problem.

### What did you do?

I attempted valiantly to construct a "composite" fetching type with associations, including has-many associations, and database functions used for bespoke column-selection.

### What did you expect to happen?

I expected to be able to do this with raw SQL.

### What happened instead?

I was not able to do this with raw SQL.

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.8.0
**Installation method:** SPM
**Xcode version:** 12.5.1
**Swift version:** 5.4.2
**Platform(s) running GRDB:** macOS
**macOS version running Xcode:** 12.0 (beta 3)

### Demo Project

For the sake of argument, suppose that I have an `Author` having-many `Book`s, each `Book` having-many `Tag`s:

```swift
struct Author: Codable, TableRecord, FetchableRecord, PersistableRecord {
    let uuid: UUID  // PRIMARY KEY
    let name: String
    let books: [Book]
}
struct Book: Codable, TableRecord, FetchableRecord, PersistableRecord {
    let uuid: UUID  // PRIMARY KEY
    let title: String
    let published: Date
    let text: String
    let author: UUID  // REFERENCES author (uuid) ON DELETE CASCADE
}
struct Tag: Codable, TableRecord, FetchableRecord, PersistableRecord {
    let name: String
    let index: Int
    let book: UUID  // REFERENCES book (uuid) ON DELETE CASCADE
}
```

This is all well and good. The GRDB documentation is superb, and it plainly describes how to create and populate a nested type using associations and joined queries and so on.

Suppose, however, that I want to be able to selectively extract columns, and not only do I want to selectively extract columns, but I also want to be able to specify the use of a database function (`f`) on `text`. In other words, I want to construct a query like this…

```sql
SELECT
    author.name AS author,
    book.published AS published,
    f(book.text) AS text,
    -- tags???
FROM author
LEFT JOIN book
    ON book.author == author.uuid
LEFT JOIN tag
    ON tag.book == book.uuid
```

…to populate a struct like this:

```swift
struct Read: Decodable, FetchableRecord {
    let author: String
    let published: Date
    let text: String
    let tags: [Tag]
}
```

This works, verbatim, for the first three variables when ensconced in an extension returning an “adapted fetch request”, as cribbed from the documentation:

```swift
extension Read {
    static func all() -> AdaptedFetchRequest<SQLRequest<Read>> {
        let request: SQLRequest<Read> = """
            SELECT
                author.uuid AS uuid,
                book.published AS published,
                f(text) AS text
                -- tags???
            FROM author
            LEFT JOIN book
                ON book.author == author.uuid
            LEFT JOIN tag
                ON tag.book == book.uuid
            """
        return request.adapted { db in
            ScopeAdapter(["author": RangeRowAdapter(0..<1),
                          "book": RangeRowAdapter(1..<3),
                          /// tags...
                         ])
        }
    }
}
```

But how can I fetch the tags as an array of `Tag` rather than as single values?

Thanks in advance.

**Comment by groue** [comment]
Hello @ManheimGreen,

There are several questions in this issue.

---

> I expected to be able to do this with raw SQL.

Fetching a list of books with their tags is not expressible with a single SQL request. This is because you can't represent a *tree* in a *table*.

What you can do is fetching a list of books, and then a list of all their tags, with *two SQL requests*, and then dispatch the fetched tags to their target books.

This dispatching task is performed for you by GRDB with the `including(all:)` association method. You can do it by hand as well. See #1005 for more information.

---

> For the sake of argument

The SQL request you show is unable to fill your `Read` struct. The `LEFT JOIN` may fetch null values, and this could crash your app when you decode non-optional properties.

Instead, we'll fill your `Read` struct with the following SQL, and an extra `including(all:)` topping:

```sql
SELECT
  f(book.text) AS text, 
  book.published, 
  author.name AS author
FROM book
JOIN author ON author.uuid = book.author
```

Look how the main table of the request is now book, which is joined (not left joined) with author, in order to avoid unwanted null values.

---

> ```swift
> struct Author: Codable, TableRecord, FetchableRecord, PersistableRecord {
>     let uuid: UUID  // PRIMARY KEY
>     let name: String
>     let books: [Book]
> }
> ```

This is not a proper GRDB record, because of the `books` property. In [Good Practices for Designing Record Types](https://github.com/groue/GRDB.swift/blob/master/Documentation/GoodPracticesForDesigningRecordTypes.md), you'll read that each record type should be responsible for its own table. There exist a `book` table, and a `Book` record. So books are not the business of `Author`. The fixed Author struct is:

```swift
struct Author: Codable, FetchableRecord, PersistableRecord {
    let uuid: UUID  // PRIMARY KEY
    let name: String
}
```

This makes you able to write, for example:

```swift
try dbQueue.write { db in
    let melville = Author(uuid: UUID(), name: "Melville")
    try melville.insert(db)
    
    let mobyDick = Book(uuid: UUID(), title: "Moby-Disk", published: Date(), text: "Call me Ishmael.", author: melville.uuid)
    try mobyDick.insert(db)
    
    try Tag(name: "Classic", index: 0, book: mobyDick.uuid).insert(db)
    try Tag(name: "Obsession", index: 1, book: mobyDick.uuid).insert(db)
    try Tag(name: "Ocean", index: 2, book: mobyDick.uuid).insert(db)
}
```

Neat!

---

## Step 1: A request that does almost what we need.

Let's first build on what the [Associations Guide](https://github.com/groue/GRDB.swift/blob/master/Documentation/AssociationsBasics.md) tells us. It tells how to fetch books with their author and their tags. From such information, we *can* build an array of `Read`, right?

```swift
// The needed associations
extension Book {
    static let author = belongsTo(Author.self)
    static let tags = hasMany(Tag.self).order(Column("index"))
}

// A temporary struct dedicated to decoding books with their author and their tags
struct Temp: Decodable, FetchableRecord {
    let book: Book
    let author: Author
    let tags: [Tag]
}

// A vanilla request that uses associations
let request = Book
    .including(required: Book.author)
    .including(all: Book.tags)
    .asRequest(of: Temp.self)

// Fetch
let temps = try request.fetchAll(db)

// Build Read
let reads = temps.map { temp in
    Read(
        author: temp.author.name,
         published: temp.book.published,
         text: temp.book.text, // missing function
         tags: temp.tags)
}
```

The code above is close to what we need.

Unfortunately:

1. We miss the function call on text
2. It triggers an internal GRDB crash due to the ambiguity of the `author` key: does it mean the `author` column of the book table, or the `author` association?
3. It fetches too much information

The crash can be solved by removing this ambiguity, as below (the association uses a different `bookAuthor` key):

```swift
struct Temp: Decodable, FetchableRecord {
    let book: Book
    let bookAuthor: Author
    let tags: [Tag]
}
let request = Book
    .including(required: Book.author.forKey("bookAuthor"))
    .including(all: Book.tags)
    .asRequest(of: Temp.self)
let temps = try request.fetchAll(db)
let reads = temps.map { temp in
    Read(author: temp.bookAuthor.name,
         published: temp.book.published,
         text: temp.book.text, // missing function
         tags: temp.tags)
}
```

## Step 2: A request that does exactly what we need.

Let's look at the `Read` record:

```swift
struct Read: Decodable, FetchableRecord {
    let author: String  // Comes from the author table
    let published: Date // Comes from the book table
    let text: String    // Comes from the book table
    let tags: [Tag]     // Comes from the tag table
}
```

The `tags` property will be fetched a second SQL request, with `including(all:)`. But the three other properties are fetched right from our base request:

```sql
SELECT
  f(book.text) AS text, 
  book.published, 
  author.name AS author
FROM book
JOIN author ON author.uuid = book.author
```

We can not use `including(all:)` with raw SQL request (see #1005). So we have to write our base request with Swift, with associations.

The record structure MUST match the [structure of a joined request](https://github.com/groue/GRDB.swift/blob/master/Documentation/AssociationsBasics.md#the-structure-of-a-joined-request), or the structure of the request must match the structure of the record, there is no escape.

The first request had us define a `Temp` record whose structure matches the request.

Now we will write the request so that its structure matches `Read`. This gives:

```swift
let author = TableAlias()                             // 1a
let request = Book
    .select(Column("uuid"),                           // 2a
            Column("text").uppercased.forKey("text"), // 3
            Column("published"),
            author[Column("name")].forKey("author"))  // 1b
    .joining(required: Book.author.aliased(author))   // 1c
    .including(all: Book.tags)                        // 2b
    .asRequest(of: Read.self)
let reads = try request.fetchAll(db)
```

1. (1a, 1b, 1c) We define a [Table Alias](https://github.com/groue/GRDB.swift/blob/master/Documentation/AssociationsBasics.md#table-aliases) so that we can select the author name at the root of the request
2. (2a, 2b) We MUST select book uuids, so that GRDB can associated tags with their proper books.
3. Here you see that we can apply a function

This request runs the following SQL request:

```sql
SELECT 
  book.uuid, 
  swiftUppercaseString(book.text) AS text, 
  book.published, 
  author.name AS author 
FROM book
JOIN author ON author.uuid = book.author;

SELECT *, book AS grdb_book
FROM tag 
WHERE book IN (...)
```

---

I hope this journey has shed some light on SQL, GRDB, and its associations!

**Comment by ManheimGreen** [comment]
> Fetching a list of books with their tags is not expressible with a single SQL request. This is because you can't represent a _tree_ in a _table_.

Good point.

> This is not a proper GRDB record, because of the `books` property. In [Good Practices for Designing Record Types](https://github.com/groue/GRDB.swift/blob/master/Documentation/GoodPracticesForDesigningRecordTypes.md), you'll read that each record type should be responsible for its own table. There exist a `book` table, and a `Book` record. So books are not the business of `Author`. The fixed `Author` struct is:

Good catch. `books` does not appear as a property of `Author` in the unabridged schema. To be clear, here is a bit of abridged table-definition code:

```
try DB.singleton.dbQueue?.write { db in
    try db.create(table: "author") { t in
        t.column("uuid", .blob).notNull().primaryKey(onConflict: .replace)
        t.column("name", .text).notNull()
    }
    try db.create(table: "book") { t in
        t.column("uuid", .blob).notNull().primaryKey(onConflict: .replace)
        t.column("published", .date).notNull()
        t.column("title", .text).notNull()
        t.column("text", .text).notNull()
        t.column("author", .blob).notNull().unique(onConflict: .replace).references("author", onDelete: .cascade)
    }
    try db.create(table: "tag") { t in
        t.column("name", .text).notNull()
        t.column("index", .integer).notNull()
        t.column("book", .blob).notNull().references("book", onDelete: .cascade)
        t.constraint(sql: "PRIMARY KEY (name, index, book) ON CONFLICT REPLACE")
    }
}
```

(Parenthetically, the value of `tag.index` is derived from `utf16Offset(in:)` — its position in `book.text`.)

And these are the associations:

```
extension Author {
    static let books = hasMany(Book.self)
    static let tags = hasMany(Tag.self, through: Author.books, using: Book.tags)
}
extension Book {
    static let author = belongsTo(Author.self)
    static let tags = hasMany(Tag.self)
}
extension Tag {
    static let book = belongsTo(Book.self)
}
```

Simple enough.

> It triggers an internal GRDB crash due to the ambiguity of the `author` key: does it mean the `author` column of the `book` table, or the `author` association?

How is this possible? `author` is a static property on `Book` — why in the world would it have anything to do with the `author` column of the `book` table?

> It fetches too much information

How so? Are you suggesting that your "vanilla request that uses associations" fetches all columns of all rows (corresponding to the request) before silently discarding some of them (the columns) when it comes time to insert ("dispatch"?) them into the `Read` struct?

> Column("text").uppercased.forKey("text")

This won't work for me because the database function I am using is one (really, several) I have defined myself. Exempli gratia:

```
static let substringHead = {
    DatabaseFunction("substringHead", argumentCount: 2, pure: true) { (values: [DatabaseValue]) in
        guard let string = String.fromDatabaseValue(values[0]),
              let upperBoundInt = Int.fromDatabaseValue(values[1])
        else { return "" }
        let upperBound = min(String.Index(utf16Offset: upperBoundInt, in: string), string.endIndex)
        guard upperBound >= string.startIndex else { return "" }
        return String(string[..<upperBound])
    }
}()
```

...Or is it possible to somehow extend `Column` with a database function?

**Comment by groue** [comment]
> How so? Are you suggesting that your "vanilla request that uses associations" fetches all columns of all rows (corresponding to the request) before silently discarding some of them (the columns) when it comes time to insert ("dispatch"?) them into the Read struct?

Yes

> This won't work for me

You didn't try `Book.select(substringHead(Column("text")), ...)`, do you?

**Comment by ManheimGreen** [comment]
The selection bit is interesting — I suppose it's in the docs somewhere, but it isn't necessarily what I would expect.

As for the database function, I had already tried `substringHead(Column("text"), 10)`, but it didn't work, which explains why I earlier missed it as a valid formulation. Upon closer inspection, the problem is not the arguments, but rather something else entirely, as is evidenced by the error emitted, this time with `reversed`:

```
GRDB/FetchableRecord+Decodable.swift:7: Fatal error: 'try!' expression unexpectedly raised an error: column not found: "text" - row: [uuid:Data(16 bytes) published:"2021-07-26 05:12:31.192" title:"Moby-Dick" reverse("book"."text"):"...leamhsI em llaC" author:"Herman Melville"]
```

So it seems that — by default — the key of the database-function-related key-value-pair somehow comes to acquire the designation (the type signature?) of the database function. Fortunately, this is easily remedied by simply appending `.forKey("text")`: `reverse(Column("text")).forKey("text")`.

I'd like to write a longer, more in-depth summary, but I don't really know where to begin. GRDB is beautiful to behold, and absolutely towers over its alleged competition, but it is often immensely magical — especially around types. Even when I know exactly what I want to do, with respect to actual implementation I spend my time more-or-less just cargo-culting the documentation with minor tweaking until it starts working.

Gwendal, you are clearly some sort of deranged Swift soothsayer-guru-man. Salut.

**Comment by groue** [comment]
GRDB suffers from the same problems as all ORMs:

- A mismatch between the Swift code and SQL (simple SQL is not always simple Swift, simple Swift is not always simple SQL, and concepts only partially overlaps)
- A huge documentation challenge, and a huge error reporting challenge, due to the combinatorial explosion of use cases

It happens that your innocuous-looking SQL request digs right into the middle of a "worst case":

- Associations API works better when it builds trees of full records (as in the first example I provided).
- Your usage of "author" as both a book column (for its author id), and a association name, is a use case that I did not expect, with unfortunate consequences and bad error reporting.

This does not mean you do anything wrong, of course. Just, the api was not tailored for this setup. My own use of the associations API usually remains 99% of the time in the simpler 1st case. It also happen that I prefer `authorId` over `author`, just so that one look tells what we are talking about.

I hope your future GRDB experiences will be easier than this one.
